# Exploring Hacker News Posts

## Introduction 

In this project, we'll work with a data set of submissions to popular technology site [Hacker News](https://news.ycombinator.com/).

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

You can find the [data set here](https://www.kaggle.com/hacker-news/hacker-news-posts), but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions. 

Below are descriptions of the columns:

`id`: The unique identifier from Hacker News for the post

`title`: The title of the post

`url`: The URL that the posts links to, if it the post has a URL

`num_points`: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes

`num_comments`: The number of comments that were made on the post

`author`: The username of the person who submitted the post

`created_at`: The date and time at which the post was submitted

We're specifically interested in posts whose titles begin with either `Ask HN` or `Show HN`. Users submit Ask HN posts to ask the Hacker News community a specific question. 

Below are a couple examples:

`Ask HN: How to improve my personal website?
Ask HN: Am I the only one outraged by Twitter shutting down share counts?
Ask HN: Aby recent changes to CSS that broke mobile?`

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. 

Below are a couple of examples:

`Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'
Show HN: Something pointless I made
Show HN: Shanhu.io, a programming playground powered by e8vm`


**We'll compare these two types of posts to determine the following:**

* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?

In [1]:
from csv import reader

In [2]:
# Read the csv file
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
hn_columns = hn[0]
hn = hn[1:]

In [3]:
# show the columns
hn_columns

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In [4]:
# show the first five rows
hn[:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

## Extracting Ask HN and Show HN Posts

In [5]:
# create three empty lists 
ask_posts = []
show_posts = []
other_posts = []

# iterate over the hn  
for row in hn:
    # save the title in the title variable
    title = row[1]
    # If the lowercase version of title starts with ask hn, append the row to ask_posts.
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    # Else if the lowercase version of title starts with show hn, append the row to show_posts.
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    # Else append to other_posts.
    else:
        other_posts.append(row)

print('The Number of ask posts: {}'.format(len(ask_posts)))
print('The Number of show posts: {}'.format(len(show_posts)))
print('The number of other posts: {}'.format(len(other_posts)))      

The Number of ask posts: 1744
The Number of show posts: 1162
The number of other posts: 17194


## Calculating the Average Number of Comments for Ask HN and Show HN Posts

In [6]:
# initialize the counter
total_ask_comments = 0 
total_show_comments = 0 

# iterate over ask posts and calculate the number of comments
for post in ask_posts:
    num_comments = float(post[4])
    total_ask_comments += num_comments 

# iterate over show posts and calculate the number of comments
for post in show_posts:
    num_comments = float(post[4])
    total_show_comments += num_comments
    
# calculate the average number of comments for ask posts 
average_ask_comments = total_ask_comments / len(ask_posts)

# calculate the average number of comments for show posts 
average_show_comments = total_show_comments / len(show_posts)

print('Average number of ask comments: {}'.format(average_ask_comments))
print('Average number of show comments: {}'.format(average_show_comments))

Average number of ask comments: 14.038417431192661
Average number of show comments: 10.31669535283993


On Average `Ask HN` posts gets more comments than the `Show HN` posts. 

## Finding the Amount of Ask Posts and Comments by Hour Created

In [7]:
# Import the datatime module
import datetime as dt 

# create an empty list 
result_list = []

# iterate over ask posts 
for post in ask_posts:
    # append the created at column value and number of comments value
    result_list.append([post[6], post[4]])
result_list[:5]

[['8/16/2016 9:55', '6'],
 ['11/22/2015 13:43', '29'],
 ['5/2/2016 10:14', '1'],
 ['8/2/2016 14:20', '3'],
 ['10/15/2015 16:38', '17']]

In [8]:
# create two empty dictionaries 
counts_by_hour = {}
comments_by_hour = {}

# iterate over the result list
for item in result_list:
    # select the date 
    date_str = item[0]
    # create a datetime object 
    date_obj = dt.datetime.strptime(date_str, "%m/%d/%Y %H:%M")
    # Extract the hour from the dattime obj
    hour = date_obj.hour
    # If the hour isn't a key in counts_by_hour
    if hour not in counts_by_hour:
        # Create the key in counts_by_hour and set it equal to 1
        counts_by_hour[hour] = 1
        # Create the key in comments_by_hour and set it equal to the comment number
        comments_by_hour[hour] = int(item[1])
        
    else: # If the hour is already a key in counts_by_hour
        # Increment the value in counts_by_hour by 1
        counts_by_hour[hour] += 1 
        # Increment the value in comments_by_hour by the comment number.
        comments_by_hour[hour] += int(item[1])
           

In [9]:
counts_by_hour

{9: 45,
 13: 85,
 10: 59,
 14: 107,
 16: 108,
 23: 68,
 12: 73,
 17: 100,
 15: 116,
 21: 109,
 20: 80,
 2: 58,
 18: 109,
 3: 54,
 5: 46,
 19: 110,
 1: 60,
 22: 71,
 8: 48,
 4: 47,
 0: 55,
 6: 44,
 7: 34,
 11: 58}

In [10]:
comments_by_hour

{9: 251,
 13: 1253,
 10: 793,
 14: 1416,
 16: 1814,
 23: 543,
 12: 687,
 17: 1146,
 15: 4477,
 21: 1745,
 20: 1722,
 2: 1381,
 18: 1439,
 3: 421,
 5: 464,
 19: 1188,
 1: 683,
 22: 479,
 8: 492,
 4: 337,
 0: 447,
 6: 397,
 7: 267,
 11: 641}

# Calculating the Average Number of Comments for Ask HN Posts by Hour

In [11]:
# create an empty list 
avg_by_hour = []

# iterate over the comments_by_hour dict
for hour in comments_by_hour:
    # append the hour and the average comments in that hour
    avg_by_hour.append([hour, round((comments_by_hour[hour] / counts_by_hour[hour]), 2)])
    
# print the result 
avg_by_hour

[[9, 5.58],
 [13, 14.74],
 [10, 13.44],
 [14, 13.23],
 [16, 16.8],
 [23, 7.99],
 [12, 9.41],
 [17, 11.46],
 [15, 38.59],
 [21, 16.01],
 [20, 21.52],
 [2, 23.81],
 [18, 13.2],
 [3, 7.8],
 [5, 10.09],
 [19, 10.8],
 [1, 11.38],
 [22, 6.75],
 [8, 10.25],
 [4, 7.17],
 [0, 8.13],
 [6, 9.02],
 [7, 7.85],
 [11, 11.05]]

## Sorting and Printing Values from a List of Lists

In [12]:
# Create an empty list
swap_avg_by_hour = []

# iterate over the avg_by_hour and swap the values and 
# append it in swap_avg_by_hour

for item in avg_by_hour:
    swap_avg_by_hour.append([item[1], item[0]])
    
# print the values 
swap_avg_by_hour

[[5.58, 9],
 [14.74, 13],
 [13.44, 10],
 [13.23, 14],
 [16.8, 16],
 [7.99, 23],
 [9.41, 12],
 [11.46, 17],
 [38.59, 15],
 [16.01, 21],
 [21.52, 20],
 [23.81, 2],
 [13.2, 18],
 [7.8, 3],
 [10.09, 5],
 [10.8, 19],
 [11.38, 1],
 [6.75, 22],
 [10.25, 8],
 [7.17, 4],
 [8.13, 0],
 [9.02, 6],
 [7.85, 7],
 [11.05, 11]]

In [13]:
# sort the swap_avg_by_hour in descending order
# and assign to sorted_swap
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

In [14]:
# print the the 5 hours with the highest average comments.
print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(str(hr), "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The `Ask HN` posts receives highest average comments at `15:00(3:00 pm est)` which is `38.59`, followed by at `2:00(2:00 am est)` which is 23.81. 

# Conclusion

Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est). 

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.